expression/builtin_time.go

```diff
@@ -2505,18 +2505,64 @@ type timeFormatFunctionClass struct {
 }
 
 func (c *timeFormatFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinTimeFormatSig{newBaseBuiltinFunc(args, ctx)}
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp.Flen = mysql.MaxBlobWidth
+	sig := &builtinTimeFormatSig{baseStringBuiltinFunc{bf}}
+	return sig.setSelf(sig), nil
 }
 
 type builtinTimeFormatSig struct {
-	baseBuiltinFunc
+	baseStringBuiltinFunc
 }
 
 // eval evals a builtinTimeFormatSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_time-format
-func (b *builtinTimeFormatSig) eval(row []types.Datum) (d types.Datum, err error) {
-	return d, errFunctionNotExists.GenByArgs("TIME_FORMAT")
+func (b *builtinTimeFormatSig) evalString(row []types.Datum) (string, bool, error) {
+	time, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if err != nil || isNull {
+		return "", isNull, errors.Trace(err)
+	}
+	formatMask, isNull, err := b.args[1].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if err != nil || isNull {
+		return "", isNull, errors.Trace(err)
+	}
+	var d types.Datum
+	d.SetString(time)
+	var res string
+	res, err = b.builtinTimeFormat(d, formatMask, b.ctx)
+	return res, isNull, err
+}
+
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_time-format
+func (b *builtinTimeFormatSig) builtinTimeFormat(time types.Datum, formatMask string, ctx context.Context) (res string, err error) {
+	var d types.Datum
+	d, err = convertToDuration(ctx.GetSessionVars().StmtCtx, time, types.MaxFsp)
```

Comment:
we should not call `convertToDuration` here, just convert a string to a duration instead

`convertToDuration` will call `func (d *Datum) ConvertTo(sc *variable.StatementContext, target *FieldType) (Datum, error)`, inside which there is `switch...case` for the target type and another `switch...case` for the origin type, which have poor performance because it do not know the origin and target type, and force cpu to do branch predicate, which may cause mis-predicate and flush pipeline and damage the performance.

all time-related functions are not refactored now, and we are working on it: https://github.com/pingcap/tidb/issues/4080

after refactoring, `convertToDuration` should be abandoned.